### DISCUSSION TITLE
Open organization links in Android Work Profile

### DISCUSSION THREAD

**Discussion by cglong** [community/community#10179] [Mobile]
My workplace enforces that I can only login using my work identity from within Android Work Profile. They also enforce that I have SAML authentication to view any of their GitHub organizations. Therefore, whenever I open a link that requires SSO, I get this:

![Screenshot_20220115-143309.png](https://user-images.githubusercontent.com/137214/149635571-e2489252-ca48-44d3-a97f-13a7c5c558cb.png)

I wish there was a button that could automatically switch to the GitHub app in my Work Profile for this link. Outlook has a similar button already, so it should be possible.

**Comment by brianlovin** [comment]
Thanks for the report here, I agree it would be great to have a faster way to switch accounts in this context. Just to clarify: are you looking for a faster way to switch accounts here? Or is the issue that your main GitHub account doesn't have the right authorization token in the first place? If it's the latter, you might be able to sign out, sign out of GitHub.com in your browser, then sign in again in the app, but make sure you tap on the green `Authorize` button next to your organizations in the sign-in flow.

**Reply by cglong** [reply]
Thank you @brianlovin! I tried logging in again, but as I feared, I'm not able to Authorize these organizations from my personal Android profile. This is enforced by Intune, which ensures that only my Work Profile is authorized to access work resources.

**EDIT:** This actually only applies to users who share a GitHub account for both personal and work use, like me. People with separate accounts for work assumedly wouldn't join their personal accounts to any organizations.

**Reply by brianlovin** [reply]
Got it, I'm not actually too familiar with what Android profiles are, or how those work. I'm wondering how much of this is in our control versus how much is going to always be restricted from the company's policies or technical restrictions on the OS. Do you have more information on how this works where I could learn?

**Reply by cglong** [reply]
Oh no worries, I appreciate you exploring this with me! Intune has made [a great intro](https://docs.microsoft.com/en-us/mem/intune/user-help/what-happens-when-you-create-a-work-profile-android) to Android Work Profiles that shows the value they bring from a UX perspective.

[A more technical description](https://developer.android.com/work/managed-profiles) is available from Android Developers. In particular, you're right about wondering about restrictions; from that article:

> The profile admin can choose which intents are allowed to cross from one profile to another. Since the IT admin makes this decision, there's no way for you to know in advance which intents are allowed to cross this boundary. The IT admin sets this policy, and is free to change it at any time.

The article shows how you can detect at runtime whether the functionality is available; I guess you could promote awareness to this intent through documentation for enterprise customers.

**Reply by cglong** [reply]
I actually just found [this example](https://developer.android.com/codelabs/work-profile-apps#5) that shows exactly how to do this! Apparently there's a `CrossProfileApps` service that you can use to switch between profiles 🥳 From the doc:

> All apps can call these APIs to launch the main activity of the same app if it's already installed in the other profile.

It seems like it only supports launching the main activity, which is still a pretty good UX since it'll work for any user with the GitHub app installed in both profiles. However, it might be worth also exploring the intent functionality to support deep-linking for any profile admin who opts in.

**Reply by cglong** [reply]
Alright, after a little more research, I think I know how this would work! I know V2 is a lot of work, but Android Work Profile is [the only way](https://developers.google.com/android/work/device-admin-deprecation) to do MDM on Android 10+, so I think this is a worthwhile investment.

### MVP
The MVP would be to implement something like [the Android Developers example](https://developer.android.com/codelabs/work-profile-apps#5) I linked above. [The `CrossProfileApps` docs verify](https://developer.android.com/reference/android/content/pm/CrossProfileApps#startMainActivity(android.content.ComponentName,%20android.os.UserHandle)) that `startMainActivity` doesn't require any permissions and can be used by any app. By adding this button to the error page above, users would be taken to the main page of the GitHub app inside their Work Profile. It's not ideal, but at least it would be a shortcut that would grant me the needed authorizations.

### V2
`CrossProfileApps` also offers a [`startActivity`](https://developer.android.com/reference/android/content/pm/CrossProfileApps#startActivity(android.content.Intent,%20android.os.UserHandle,%20android.app.Activity)) method, which requires the `INTERACT_ACROSS_PROFILES` permission. From a quick search, it seems that, to publish an app in Google Play that has this permission, you need to apply for an [Early Access Program](https://developers.google.com/android/work/connected-apps#early_access_program). Therefore, I'd recommend this plan:

1. Download and start integrating [the connected apps library](https://developers.google.com/android/work/connected-apps#getting_started).
2. In the code, check for the following. If any of the following return `false`, fallback to the MVP UX:
    a. Has the `INTERACT_ACROSS_PROFILES` permission been granted?
    b. Has the profile admin enabled both versions of the app to connect to one another?
    c. Has the user approved the apps to connect?
    d. Does the work version of the app have the needed token access?
3. If all of the above are `true`, somehow open the current resource with the right token access, e.g. call `CrossProfileApps.startActivity` to open it in the work profile.
4. Finally, complete and submit the [connected apps questionnaire](https://developers.google.com/android/work/connected-apps#requesting_approval).

**Reply by aconsuegra** [reply]
Thanks for your details answer @cglong! This will be super useful for our engineers. Do you mind sharing a screenshot here about how the Outlook app deals with this flow?

**Reply by cglong** [reply]
Yes for sure @aconsuegra; I'm happy to help however I can 🙂 The Outlook app has this flow embedded as part of its multi-account flow:

1. Here's the default view when I use Outlook in my personal profile:
![Screenshot_20220124-203305](https://user-images.githubusercontent.com/137214/151106263-51367a4e-0c8d-48c8-8f0b-d5e98f89a198.png)
2. When I tap the accounts picker in the top-left, I see all my accounts, including a briefcase icon:
![Screenshot_20220124-203311](https://user-images.githubusercontent.com/137214/151106299-4f666d19-0cf5-4e0f-a077-76371e9aa1a9.png)
3. When I tap the briefcase, I'm taken to the default view in my work profile: 
![Screenshot_20220124-203443](https://user-images.githubusercontent.com/137214/151106372-176955cd-bd20-4dc4-aa3d-2517d917e29e.png)
4. If I open the accounts picker again, I see a person icon, which will take me back to my personal profile: 
![Screenshot_20220124-203431](https://user-images.githubusercontent.com/137214/151106404-b826c615-5d8b-459b-a13d-ad6bed78923d.png)

Thank you again for looking into this with me! Please let me know if there's anything else I can do to help out here 🙂

**Comment by Mobileinstallment** [comment]
38000
